### PR TITLE
feat: enable BYTES for LPAD and RPAD

### PR DESCRIPTION
**Pull request by Sullivan-Patrick**
### Description 
This PR enables LPAD and RPAD UDFs to accept BYTES as a parameter

### Testing done 
Unit testing and QTTs

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 1]
```diff
@@ -987,9 +987,10 @@ Since: 0.10.0
 RPAD(input, length, padding)
 ```
 
-Pads the input string, starting from the end, with the specified padding string until the target length is reached. If the input string is longer than the specified target length it will be truncated. 
+Pads the input string or bytes, starting from the end, with the specified padding of the same type until the target length is reached. 
+If the input is longer than the specified target length, it is truncated.
```
Comment (Sullivan-Patrick) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:991]: This is just GitHub formatting making spacing look inconsistent, it's consistent.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java [hunk 2]
```diff
@@ -49,4 +52,41 @@ public String rpad(
...
+      return null;
+    }
+    if (padding == null
+        || BytesUtils.getByteArray(padding).length == 0
+        || targetLen == null
+        || targetLen < 0) {
+      return null;
+    }
+
+    final byte[] start = BytesUtils.getByteArray(input);
```
Comment (Sullivan-Patrick) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java:72]: I think working with a `byte[]` is a lot more readable than trying to work with a buffer here, do others agree/disagree?

Comment (spena) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java:72]: Yes. I'm actually working in a PR to move away from ByteBuffer because some issues with internal positions of it.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java [hunk 3]
```diff
@@ -49,4 +52,41 @@ public String rpad(
...
+  public ByteBuffer rpad(
+      @UdfParameter(description = "Bytes to be padded") final ByteBuffer input,
+      @UdfParameter(description = "Target length") final Integer targetLen,
+      @UdfParameter(description = "Padding bytes") final ByteBuffer padding) {
+
+    if (input == null) {
+      return null;
+    }
+    if (padding == null
+        || BytesUtils.getByteArray(padding).length == 0
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java:66]: Same comment regarding calling `getByteArray` once because full copies when using protobuf bytes.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java [hunk 4]
```diff
@@ -49,4 +52,41 @@ public String rpad(
...
+
+    for (int i = 0; i < targetLen; i++) {
+      if (i < start.length) {
+        padded[i] = start[i];
+      } else {
+        final int paddingIndex = (i - start.length) % paddingArray.length;
+        padded[i] = paddingArray[paddingIndex];
+      }
+
+    }
```
Comment (spena) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java:89]: You could start copying the initial bytes by using `Sytem.arraycopy`, and then just adding the padding bytes.

```
// This copies all bytes from `start` to `padded`.
// See https://www.tutorialspoint.com/java/lang/system_arraycopy.htm
System.arraycopy(start, 0, padded, 0, start.length);
```

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/LPadTest.java [hunk 5]
```diff
@@ -18,63 +18,130 @@
...
+    return ByteBuffer.wrap(new byte[]{1,2,3});
+  }
+
+  private ByteBuffer padding() {
+    return ByteBuffer.wrap(new byte[]{4,5});
+  }
+
+  private ByteBuffer empty(){
+    return ByteBuffer.wrap(new byte[]{});
+  }
```
[the remark this answers is not in the document]
Comment (Sullivan-Patrick) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/LPadTest.java:145]: I was trying to ensure that the tests were idempotent. I suppose that doesn't really matter given the test file is small and we clearly aren't changing the value of the arrays?

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/LPad.java [hunk 6]
```diff
@@ -50,4 +53,42 @@ public String lpad(
...
+      return null;
+    }
+    if (padding == null
+        || BytesUtils.getByteArray(padding).length == 0
+        || targetLen == null
+        || targetLen < 0) {
+      return null;
+    }
+
+    final byte[] start = BytesUtils.getByteArray(input);
```
[the remark this answers is not in the document]
Comment (Sullivan-Patrick) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/LPad.java:73]: I believe I'm only calling this once, do you mean `ByteUtils.getByteArray(padding)`? Or am I missing something?

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java (hunk 4 above)
Comment (Sullivan-Patrick) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/RPad.java:89]: Good idea, this is cleaner.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/LPadTest.java [hunk 7]
```diff
@@ -32,10 +33,21 @@ public void shouldPadInputString() {
 
   @Test
   public void shouldPadInputBytes() {
-    final ByteBuffer result = udf.lpad(input(), 7, padding());
+    final ByteBuffer result = udf.lpad(BYTES_123, 7, BYTES_45);
     assertThat(result, is(ByteBuffer.wrap(new byte[]{4,5,4,5,1,2,3})));
   }
 
+  @Test
```
Comment (Sullivan-Patrick) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/LPadTest.java:40]: Original author had added this test on rpad but not rpad, just noticed and added them here.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/LPadTest.java [hunk 8]
```diff
@@ -32,10 +33,21 @@ public void shouldPadInputString() {
...
+  public void shouldAppendPartialPaddingString() {
+    final String result = udf.lpad("foo", 4, "Bar");
+    assertThat(result, is("Bfoo"));
+  }
+
+  @Test public void shouldAppendPartialPaddingBytes() {
+    final ByteBuffer result = udf.lpad(BYTES_123, 4, BYTES_45);
+    assertThat(BytesUtils.getByteArray(result), is(new byte[]{4,1,2,3}));
+  }
+
```
Comment (Sullivan-Patrick) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/LPadTest.java:50]: Original author had added this test on rpad but not rpad, just noticed and added them here.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/LPad.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (spena) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/string/LPad.java:73]: Got confused. It's good then.

**Review by spena — APPROVED** [review]
+1

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java [hunk 9]
```diff
@@ -18,69 +18,134 @@
...
+    assertThat(BytesUtils.getByteArray(result), is(new byte[]{1,2,3,4,5,4,5}));
+  }
+
+  @Test
+  public void shouldAppendPartialPaddingString() {
     final String result = udf.rpad("foo", 4, "Bar");
     assertThat(result, is("fooB"));
   }
 
+  @Test public void shouldAppendPartialPaddingBytes() {
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java:46]: nit: newline.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java [hunk 10]
```diff
@@ -18,69 +18,134 @@
...
+  public void shouldPadInputString() {
     final String result = udf.rpad("foo", 7, "Bar");
     assertThat(result, is("fooBarB"));
   }
 
   @Test
-  public void shouldAppendPartialPadding() {
+  public void shouldPadInputBytes() {
+    final ByteBuffer result = udf.rpad(BYTES_123, 7, BYTES_45);
+    assertThat(BytesUtils.getByteArray(result), is(new byte[]{1,2,3,4,5,4,5}));
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java:37]: The new unit tests for bytes don't seem to cover the case where target length is greater than start length and the difference is not an integer multiple of the length of the padding array.

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java [hunk 11]
```diff
@@ -18,69 +18,134 @@
...
     assertThat(result, is(nullValue()));
   }
 
+  @Test
+  public void shouldReturnNullForNullLengthBytes() {
+    final ByteBuffer result = udf.rpad(BYTES_123, null, BYTES_45);
+    assertThat(result, is(nullValue()));
+  }
+
+  private final ByteBuffer BYTES_123 = ByteBuffer.wrap(new byte[]{1,2,3});
```
Comment (vcrfxia) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java:147]: nit: let's keep the member variables together. We can move these to the top of the file where there's already
```
private final RPad udf = new RPad();
```
Also, let's make these static.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/bytes-lpad-rpad.json [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        {"topic": "test_topic", "key": "r2", "value": {"subject": "eWVz", "len": 5, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r3", "value": {"subject": "eWVz", "len": 3, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r4", "value": {"subject": "eWVz", "len": 2, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r5", "value": {"subject": "eWVz", "len": 0, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r6", "value": {"subject": "eWVz", "len": -1, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r7", "value": {"subject": "eWVz", "len": 3, "padding": ""}},
+        {"topic": "test_topic", "key": "r8", "value": {"subject": null, "len": 3, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r9", "value": {"subject": "eWVz", "len": null, "padding": "bm8="}},
+        {"topic": "test_topic", "key": "r10", "value": {"subject": "eWVz", "len": -1, "padding": null}}
+
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/bytes-lpad-rpad.json:56]: super nit: unnecessary whitespace.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/bytes.json [hunk 13]
```diff
@@ -1,5 +1,5 @@
 {
-  "comments": ["tests for decimal functionality"],
+  "comments": ["tests for bytes functionality"],
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/bytes.json:2]: Nice fix :)

**Review by vcrfxia — APPROVED** [review]
Thanks @Sullivan-Patrick -- awesome stuff! A bunch of nits inline. Great test coverage!

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java (hunk 10 above)
Comment (Sullivan-Patrick) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java:37]: Doesn't `shouldAppendPartialPaddingBytes()` do this?

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 14]
```diff
@@ -987,9 +987,10 @@ Since: 0.10.0
 RPAD(input, length, padding)
 ```
 
-Pads the input string, starting from the end, with the specified padding string until the target length is reached. If the input string is longer than the specified target length it will be truncated. 
+Pads the input string or bytes, starting from the end, with the specified padding of the same type until the target length is reached. 
+If the input is longer than the specified target length, it is truncated. 
 
-If the padding string is empty or NULL, or the target length is negative, then NULL is returned.
+If the padding string or bytes is empty or NULL, or the target length is negative, NULL is returned.
```
Comment (JimGalasyn) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:993]:
```suggestion
If the padding string or byte array is empty or NULL, or the target length is negative, NULL is returned.
```

**Review by JimGalasyn — APPROVED** [review]
LGTM, with one copy edit.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/bytes.json (hunk 13 above)
Comment (Sullivan-Patrick) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/bytes.json:2]: what would the ksql team do without me? :')

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java (hunk 10 above)
Comment (vcrfxia) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/string/RPadTest.java:40]: Sorry my previous comment was incomplete -- the missing test coverage is the case where the padding needs to wrap around (pad multiple occurrences) but not an integer number of times. It's a minor point, though. Let's merge now to get this in!
